File: python/src/cosmoner/secrets.py
"""Secrets service namespace — encrypted configuration, revealed only as it is set."""

from __future__ import annotations

import re
from typing import Any

from ._config import ClientConfig, resolve_project_id
from ._transport import AsyncTransport, Transport

_NAME_PATTERN = re.compile(r"^[A-Z][A-Z0-9_]*$")
_MAX_NAME_LENGTH = 100
_MAX_VALUE_LENGTH = 10_000


def _require_secret_id(secret_id: str) -> None:
    """Rejects an empty secret id before it becomes a malformed route."""
    if not secret_id:
        raise ValueError("secret_id is required")


def _require_name(name: str) -> None:
    """Rejects a name the API would reject, using the message the API would send."""
    if not name:
        raise ValueError("name is required")
    if not _NAME_PATTERN.match(name):
        raise ValueError(
            "Name must be uppercase letters, numbers, or underscores, "
            "and start with a letter (e.g. DB_PASSWORD)"
        )
    if len(name) > _MAX_NAME_LENGTH:
        raise ValueError(f"name must be at most {_MAX_NAME_LENGTH} characters")


def _require_value(value: str) -> None:
    """Rejects a value the API would reject."""
    if not value:
        raise ValueError("value is required")
    if len(value) > _MAX_VALUE_LENGTH:
        raise ValueError(f"value must be at most {_MAX_VALUE_LENGTH} characters")


def _environment_params(environment: str | None) -> dict[str, str] | None:
    """Builds the query that scopes a list to one environment, or none for all."""
    return {"environment": environment} if environment else None


def _write_body(
    value: str,
    description: str | None,
    name: str | None = None,
    environment: str | None = None,
) -> dict[str, Any]:
    """Builds a create or update body, omitting what the caller left out."""
    body: dict[str, Any] = {}
    if name is not None:
        body["name"] = name
    body["value"] = value
    if description is not None:
        body["description"] = description
    if environment is not None:
        body["environment"] = environment
    return body


class SecretsService:
    """Synchronous management of a project's secrets.

    Reads need the ``secrets:read`` scope. Writes need ``secrets:write`` *and*
    an owner or admin: the API checks the member's role independently of the
    key's scopes, so a plain member's key is refused even when it carries the
    scope.

    A secret's value is returned exactly once, by the call that sets it. There
    is no route that decrypts one afterwards.
    """

    def __init__(self, transport: Transport, config: ClientConfig) -> None:
        """Binds the namespace to the client's transport and resolved configuration."""
        self._transport = transport
        self._config = config

    def _base_path(self, project_id: str | None) -> str:
        """Builds the collection route for the resolved project."""
        project = resolve_project_id(self._config, project_id)
        return f"/v1/projects/{project}/secrets"

    def list(
        self, *, environment: str | None = None, project_id: str | None = None
    ) -> dict[str, Any]:
        """Lists the project's secrets as metadata. Values are never included.

        Args:
            environment: One of ``default``, ``development``, ``staging`` or
                ``production``. Omit to list every environment.
            project_id: Overrides the client-level default project.
        """
        result: dict[str, Any] = self._transport.request(
            "GET",
            self._base_path(project_id),
            params=_environment_params(environment),
        )
        return result

    def get(self, secret_id: str, *, project_id: str | None = None) -> dict[str, Any]:
        """Fetches one secret's metadata. The value is not part of the response."""
        _require_secret_id(secret_id)

        result: dict[str, Any] = self._transport.request(
            "GET", f"{self._base_path(project_id)}/{secret_id}"
        )
        return result

    def create(
        self,
        name: str,
        value: str,
        *,
        description: str | None = None,
        environment: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Stores a new secret, and returns its plaintext value once.

        The response carries ``value`` and ``maskedValue``; a later read gives
        metadata only, so store the value now rather than expect to recover it.

        Creation is rate-limited to 10 requests per 10 minutes, and a project at
        its secret limit answers 402 — see :meth:`usage`.
        """
        _require_name(name)
        _require_value(value)

        result: dict[str, Any] = self._transport.request(
            "POST",
            self._base_path(project_id),
            json=_write_body(value, description, name=name, environment=environment),
        )
        return result

    def update(
        self,
        secret_id: str,
        value: str,
        *,
        description: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Replaces the value, bumps the version, and returns the new value once."""
        _require_secret_id(secret_id)
        _require_value(value)

        result: dict[str, Any] = self._transport.request(
            "PATCH",
            f"{self._base_path(project_id)}/{secret_id}",
            json=_write_body(value, description),
        )
        return result

    def delete(self, secret_id: str, *, project_id: str | None = None) -> None:
        """Permanently removes a secret. The API answers 204, so nothing is returned."""
        _require_secret_id(secret_id)

        self._transport.request("DELETE", f"{self._base_path(project_id)}/{secret_id}")

    def usage(self, *, project_id: str | None = None) -> dict[str, Any]:
        """Reports how many secrets the project holds, may hold, and the price of more."""
        result: dict[str, Any] = self._transport.request(
            "GET", f"{self._base_path(project_id)}/usage"
        )
        return result

    def audit(self, secret_id: str, *, project_id: str | None = None) -> dict[str, Any]:
        """Reads a secret's audit trail: who changed it and when, never to what."""
        _require_secret_id(secret_id)

        result: dict[str, Any] = self._transport.request(
            "GET", f"{self._base_path(project_id)}/{secret_id}/audit"
        )
        return result


class AsyncSecretsService:
    """Asynchronous counterpart to :class:`SecretsService`."""

    def __init__(self, transport: AsyncTransport, config: ClientConfig) -> None:
        """Binds the namespace to the client's transport and resolved configuration."""
        self._transport = transport
        self._config = config

    def _base_path(self, project_id: str | None) -> str:
        """Builds the collection route for the resolved project."""
        project = resolve_project_id(self._config, project_id)
        return f"/v1/projects/{project}/secrets"

    async def list(
        self, *, environment: str | None = None, project_id: str | None = None
    ) -> dict[str, Any]:
        """Lists the project's secrets as metadata. Values are never included."""
        result: dict[str, Any] = await self._transport.request(
            "GET",
            self._base_path(project_id),
            params=_environment_params(environment),
        )
        return result

    async def get(
        self, secret_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Fetches one secret's metadata. The value is not part of the response."""
        _require_secret_id(secret_id)

        result: dict[str, Any] = await self._transport.request(
            "GET", f"{self._base_path(project_id)}/{secret_id}"
        )
        return result

    async def create(
        self,
        name: str,
        value: str,
        *,
        description: str | None = None,
        environment: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Stores a new secret, and returns its plaintext value once."""
        _require_name(name)
        _require_value(value)

        result: dict[str, Any] = await self._transport.request(
            "POST",
            self._base_path(project_id),
            json=_write_body(value, description, name=name, environment=environment),
        )
        return result

    async def update(
        self,
        secret_id: str,
        value: str,
        *,
        description: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Replaces the value, bumps the version, and returns the new value once."""
        _require_secret_id(secret_id)
        _require_value(value)

        result: dict[str, Any] = await self._transport.request(
            "PATCH",
            f"{self._base_path(project_id)}/{secret_id}",
            json=_write_body(value, description),
        )
        return result

    async def delete(self, secret_id: str, *, project_id: str | None = None) -> None:
        """Permanently removes a secret. The API answers 204, so nothing is returned."""
        _require_secret_id(secret_id)

        await self._transport.request(
            "DELETE", f"{self._base_path(project_id)}/{secret_id}"
        )

    async def usage(self, *, project_id: str | None = None) -> dict[str, Any]:
        """Reports how many secrets the project holds, may hold, and the price of more."""
        result: dict[str, Any] = await self._transport.request(
            "GET", f"{self._base_path(project_id)}/usage"
        )
        return result

    async def audit(
        self, secret_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Reads a secret's audit trail: who changed it and when, never to what."""
        _require_secret_id(secret_id)

        result: dict[str, Any] = await self._transport.request(
            "GET", f"{self._base_path(project_id)}/{secret_id}/audit"
        )
        return result

File: javascript/src/services/secrets.ts
/* eslint-disable require-await -- Every method here validates its arguments before
   reaching the transport. Keeping them `async` makes a bad argument reject the
   returned promise rather than throw synchronously, so one `.catch()` covers both
   client-side and server-side failures. */

/** Secrets service namespace — encrypted project configuration, revealed only as it is set. */

import { resolveProjectId, type ResolvedConfig } from "../config";
import type { Transport } from "../transport";
import {
  requireEntryName,
  requireEntryValue,
  type Actor,
  type ProjectEnvironment,
} from "./project-config";

/**
 * A secret's metadata.
 *
 * The value is deliberately absent: it is encrypted at rest and returned
 * exactly once, by the call that sets it. Reads describe the secret without
 * ever handing the plaintext back, and no API route decrypts one.
 */
export interface ProjectSecret {
  id: string;
  /** Uppercase, underscore-separated, e.g. `DB_PASSWORD`. */
  name: string;
  description: string | null;
  environment: ProjectEnvironment;
  /** Incremented every time the value is replaced. Starts at 1. */
  version: number;
  /** Id of the member who created it. */
  createdBy: string;
  /** Id of the member who last changed it. */
  updatedBy: string;
  createdByUser: Actor;
  updatedByUser: Actor;
  createdAt: string;
  updatedAt: string;
}

/**
 * A secret together with its plaintext value, returned once at the moment it
 * is set.
 *
 * There is no way to read the value back afterwards, so a caller that discards
 * this response has to set a new value rather than recover this one.
 */
export interface RevealedProjectSecret {
  id: string;
  name: string;
  description: string | null;
  environment: ProjectEnvironment;
  version: number;
  /** Present on create. */
  createdAt?: string;
  /** Present on update. */
  updatedAt?: string;
  /** The plaintext value. Returned only here — store it before discarding the response. */
  value: string;
  /** The value with its middle replaced, e.g. `su••••et`. Safe to display. */
  maskedValue: string;
}

/** Secret capacity and what it costs to add more. */
export interface SecretsUsage {
  /** Secrets currently stored on the project. */
  used: number;
  /** How many the project may store, free allowance plus purchased packs. */
  limit: number;
  /** Secrets included before any pack is bought. */
  freeLimit: number;
  /** How many secrets one pack adds. */
  packSize: number;
  /** Packs currently on the subscription. */
  paidPacks: number;
  /** Null when pack pricing is unavailable. */
  packPrice: { monthly: number; currency: string } | null;
}

/** One entry in a secret's audit trail. Values never appear here — only who did what, and when. */
export interface SecretAuditEntry {
  id: string;
  secretId: string;
  action: "CREATED" | "UPDATED" | "DELETED";
  actorId: string;
  actor: Actor;
  metadata: string | null;
  createdAt: string;
}

/** Options accepted by every method, for working across projects. */
export interface ProjectScopedParams {
  /** Overrides the client-level default project for this call. */
  projectId?: string;
}

/** Arguments accepted by `client.secrets.list()`. */
export interface ListSecretsParams extends ProjectScopedParams {
  /** Limits the list to one environment. Omit to list them all. */
  environment?: ProjectEnvironment;
}

/** Arguments accepted by `client.secrets.create()`. */
export interface CreateSecretParams extends ProjectScopedParams {
  /** Uppercase, underscore-separated, e.g. `DB_PASSWORD`. */
  name: string;
  value: string;
  description?: string;
  /** Defaults to `default` server-side. */
  environment?: ProjectEnvironment;
}

/** Arguments accepted by `client.secrets.update()`. */
export interface UpdateSecretParams extends ProjectScopedParams {
  /** Replaces the stored value and bumps `version`. */
  value: string;
  description?: string;
}

interface Envelope<T> {
  success: true;
  data: T;
}

export type ListSecretsResponse = Envelope<ProjectSecret[]>;
export type GetSecretResponse = Envelope<ProjectSecret>;
export type SetSecretResponse = Envelope<RevealedProjectSecret>;
export type GetSecretsUsageResponse = Envelope<SecretsUsage>;
export type GetSecretAuditResponse = Envelope<SecretAuditEntry[]>;

/**
 * Manages a project's secrets.
 *
 * Reads need `secrets:read`. Writes need `secrets:write` *and* an owner or
 * admin: the API checks the member's role independently of the key's scopes,
 * so a plain member's key is refused even when it carries the scope.
 */
export class SecretsService {
  constructor(
    private readonly transport: Transport,
    private readonly config: ResolvedConfig
  ) {}

  /** Builds the collection route for the resolved project. */
  private basePath(projectId?: string): string {
    return `/v1/projects/${resolveProjectId(this.config, projectId)}/secrets`;
  }

  /** Lists the project's secrets as metadata. Values are never included. */
  async list(params: ListSecretsParams = {}): Promise<ListSecretsResponse> {
    return this.transport.request<ListSecretsResponse>("GET", this.basePath(params.projectId), {
      query: { environment: params.environment },
    });
  }

  /** Fetches one secret's metadata. The value is not part of the response. */
  async get(secretId: string, params: ProjectScopedParams = {}): Promise<GetSecretResponse> {
    if (!secretId) throw new Error("secretId is required");

    return this.transport.request<GetSecretResponse>(
      "GET",
      `${this.basePath(params.projectId)}/${secretId}`
    );
  }

  /**
   * Stores a new secret, and returns its plaintext value once.
   *
   * Creation is rate-limited to 10 requests per 10 minutes, so a loop that
   * imports many secrets will meet a `RateLimitError`. A project at its limit
   * answers 402 — see `usage()`.
   */
  async create(params: CreateSecretParams): Promise<SetSecretResponse> {
    requireEntryName(params.name);
    requireEntryValue(params.value);

    return this.transport.request<SetSecretResponse>("POST", this.basePath(params.projectId), {
      body: {
        name: params.name,
        value: params.value,
        description: params.description,
        environment: params.environment,
      },
    });
  }

  /** Replaces a secret's value, bumps its version, and returns the new value once. */
  async update(secretId: string, params: UpdateSecretParams): Promise<SetSecretResponse> {
    if (!secretId) throw new Error("secretId is required");
    requireEntryValue(params.value);

    return this.transport.request<SetSecretResponse>(
      "PATCH",
      `${this.basePath(params.projectId)}/${secretId}`,
      { body: { value: params.value, description: params.description } }
    );
  }

  /** Permanently removes a secret. The API answers 204, so there is nothing to return. */
  async delete(secretId: string, params: ProjectScopedParams = {}): Promise<void> {
    if (!secretId) throw new Error("secretId is required");

    await this.transport.request<void>(
      "DELETE",
      `${this.basePath(params.projectId)}/${secretId}`
    );
  }

  /** Reports how many secrets the project holds, how many it may hold, and the price of more. */
  async usage(params: ProjectScopedParams = {}): Promise<GetSecretsUsageResponse> {
    return this.transport.request<GetSecretsUsageResponse>(
      "GET",
      `${this.basePath(params.projectId)}/usage`
    );
  }

  /** Reads a secret's audit trail: who changed it and when, never what it was changed to. */
  async audit(secretId: string, params: ProjectScopedParams = {}): Promise<GetSecretAuditResponse> {
    if (!secretId) throw new Error("secretId is required");

    return this.transport.request<GetSecretAuditResponse>(
      "GET",
      `${this.basePath(params.projectId)}/${secretId}/audit`
    );
  }
}
